### ISSUE TITLE
distinguishing multi-allele genotypes

### ISSUE THREAD

**Issue by ValWood** [pombase/canto#1520]
distinguising multi -allele genotypes is not as easy as it should be.
How can this be improved?
I don't think its incredibly obvious that the alleles are distinguished by the shading.

![multi](https://user-images.githubusercontent.com/7359272/41879783-062f2b6e-78d3-11e8-9bde-e8cf3337ae70.jpg)

**Comment by jseager7** [comment]
**Option 1: Merge allele rows**

Maybe we could merge the rows that correspond to a multi-allele genotype, like we already do for the multi-allele annotations summary?

![merged_annotation](https://user-images.githubusercontent.com/37659591/41965947-c09e6442-79f5-11e8-8b4c-3cad16b12a12.PNG)

Or we could at least make them look like they're merged by removing some of the inner borders.

- - -

**Option 2: Don't use tables**

This option is more controversial, and would take significantly more work: we'd move away from using a tabular style and split the genotypes into their own 'boxes', as shown in the latest mock-ups:

![compact_alleles](https://user-images.githubusercontent.com/37659591/41966062-0a07af4e-79f6-11e8-9a19-83290712a538.png)

![single_mutant_exp](https://user-images.githubusercontent.com/37659591/41966063-0c614cdc-79f6-11e8-9e1c-b382148b5c29.png)

This is probably the clearest way to distinguish each genotype, but it also takes up a lot more vertical space compared to the table, and could be difficult for users to 'scan' over.

**Comment by ValWood** [comment]
AI Val will add a mockup  (option 3) of a small padding between genotpes

**Comment by ValWood** [comment]
a bit like this...

![table_mockup](https://user-images.githubusercontent.com/7359272/42042118-55c19288-7aeb-11e8-9cd0-6b4096c16497.jpg)

**Comment by jseager7** [comment]
Option 3 would probably be fairly easy to achieve, and it resembles Option 2 in the way it visually separates the table rows. I guess there's also the advantage that you can still discern the individual alleles in the genotype. @kimrutherford, if you agree, would you like me to try and test that row padding in Canto?

**Comment by ValWood** [comment]
One problem with option 3 is that it wouldn't match the current colour alternation elsewhere. do we need to think about this first?

**Comment by mah11** [comment]
Could the padding be deployed without changing the color alternation? If so, problem evaded ;)

**Comment by ValWood** [comment]
Could try.
I think it could work, combined with lighter lines within the allele (or even no lines?)

**Comment by ValWood** [comment]
@jseager7  wait for @kimrutherford   though in case there are any catches

**Comment by kimrutherford** [comment]
> @jseager7 wait for @kimrutherford though in case there are any catches

I can't think of any cathes.  If you're happy I think we should try it.

Thanks James.

**Comment by jseager7** [comment]
Ok, so this a lot more complicated than I thought, mostly because browsers are very stubborn about styling table rows. It's not possible to apply padding directly to the rows; the best I've managed so far is a pretty ungraceful hack where empty text is added before the rows and then padding is applied to that instead, but that breaks the border effect for selected rows, and causes problems with the row backgrounds:

![row_spacing_issues](https://user-images.githubusercontent.com/37659591/42274193-af1de894-7f83-11e8-9a68-ad51dec334d0.png)

At least this shows that it's theoretically possible to add padding without changing the colour alternation.

There are a few other options I've found to fix this, but none of them are ideal:

* Option 1
  * **Solution**: Add empty rows to the markup between the existing rows, and add padding to them.
  * **Problems**: Causes accessibility issues (screen readers will report the wrong number of rows); the alternating background colours will almost certainly break; and copying and pasting the table will, in theory, paste the empty rows as well (though I'm not sure how likely it is that users will want to take a copy of the genotype tables).

* Option 2
  * **Solution**: Don't use table elements for the markup &ndash; use generic containing elements (`<div>` and `<span>`) and style them to _look_ like tables. This allows full use of style properties such as `margin` and `padding`.
  * **Problems:** Requires re-writing all the table templates; loses the semantics of the table markup, which again could cause accessibility problems because the screen reader might not know it's a table; also loses all of the table styling that we got 'for free' using the current layout.

- - -

@kimrutherford I think this task is being complicated by the way the templates are designed at the moment: for example, the `ng-repeat` logic causes every table row to be wrapped in a `<tbody>` element, rather than just being a `<tr>`, which isn't a conventional use for `<tbody>` (although I presume the reason was because it was the only way to support the alternating colours and grouping for multi-allele genotypes). I might need your help to make sense of how the templates fit together.

**Comment by jseager7** [comment]
@kimrutherford Just now I thought the problem with the selection border and the background colour could be solved if those CSS selectors targeted the `<tr>` elements _inside_ the `<tbody>` rather than the `<tbody>` element itself.

While that would work for single allele genotypes, it doesn't help us when it comes to multi-allele genotypes, because we want the selection border to cover a group of rows, not just one. I thought about nesting `<tbody>` elements to handle this, but that's a violation of the CSS specification (`<tbody>` only permits `<tr>` as children), so browsers can just choose to ignore it.

I'll have to keep thinking.

**Comment by kimrutherford** [comment]
> Don't use table elements for the markup – use generic containing elements (<div> and <span>) and
> style them to look like tables.

That would be OK as long as we can get the columns to line up nicely.

> although I presume the reason was because it was the only way to support the alternating colours and
> grouping for multi-allele genotypes

Yep.  Multi-allele genotypes complicate things.  Also, using <tbody> allowed creating a directive that could render a single genotype rather than having all the logic in one big template.

I can't think of a good solution for adding a gap.  Let's have chat about it sometime.

**Comment by ValWood** [comment]
This isn't super important right now.
We might think of another way....

If it is easy, making the lines within the allele table disappear  might help. If difficult we can continue to think about a better solution. This is only an enhancement so it isn't high priority.

**Comment by jseager7** [comment]
> making the lines within the allele table disappear might help.

Ironically, this is actually easier to do than adding gaps:

![multi-allele_no_borders](https://user-images.githubusercontent.com/37659591/42276642-6f2dc3fe-7f8c-11e8-81c3-cde4ad974c24.png)

There's one small problem with borders being added back in near the tick boxes on the single allele rows, but that's just because of the order I defined the styling rules &ndash; shouldn't be too difficult to fix.

**Comment by ValWood** [comment]
That might be a better way to go
@mah11 @Antonialock ?

**Comment by jseager7** [comment]
> Is it also worth trying faint lines or dashed lines between the alleles?

That might be more difficult because as far as I know there's no way for the CSS selectors to distinguish the multi-allele genotype table &ndash; not by class, anyway &ndash; so the single allele rows would get the same borders.

The current solution is basically a workaround involving CSS pseudo-selector trickery:

```css
/* Unset all the top and bottom borders by default */
.curs-genotype-list > tbody > tr > td {
  border-top: none;
  border-bottom: none;
}

/* Set a top border only for the first row in a genotype */
.curs-genotype-list > tbody > tr:first-child > td {
  border-top: 1px solid #bbb;
}

/*  Set a bottom border only for the last row in a genotype */
/*  Also handle the case where a cell spans multiple rows (e.g. annotations) */
.curs-genotype-list > tbody > tr:last-child > td,
.curs-genotype-list > tbody > tr > td[rowspan] {
  border-bottom: 1px solid #bbb;
}
```

**Comment by ValWood** [comment]
@Antonialock  @mah11  can you let us know what you think about this when convenient. It isn't super urgent but @jseager7  has it on hold pending our feedback.

It would be a nice quick task to slot in.

**Comment by mah11** [comment]
I'm not the one who needed any changes from the currently-live display. But I wouldn't have a problem with removing the lines between alleles *if* there's no wrapping of allele descriptions to throw off the alignment between gene and allele.

**Comment by Antonialock** [comment]
Whoa, long ticket!

I don’t understand the ‘there is a problem with borders being added back
in’




-- 
Antonia Lock, PhD
PomBase Biocurator, http://www.pombase.org
Department of Genetics, Evolution and Environment,
The Darwin Building,
University College London
London WC1E 6BT, UK

**Comment by ValWood** [comment]
It's just a constraint of the html for table drawing.

@jseager7  if you implement  what you discussed with @kimrutherford  in the final few entries I'm sure that will be fine, but If it doesn't work it can stay as it is now.

**Comment by jseager7** [comment]
> I wouldn't have a problem with removing the lines between alleles if there's no wrapping of allele descriptions

@mah11 Unfortunately, wrapping does happen if there are spaces in the description, and it does ruin the alignment of gene names and descriptions (please ignore the nonsensical data):

![wrapping-no-borders](https://user-images.githubusercontent.com/37659591/43583476-84b461e2-9657-11e8-82af-e08bdaa59600.PNG)

So I guess we're going to have to come up with some way to fix that before we can remove the lines. Maybe we could truncate the allele descriptions after a certain length like @ValWood mentioned in #1587, but we should bear the following problem in mind:

> This [truncating descriptions] might be a little more controversial because you need to pay more attention to the variations here. However, I think it *might * be OK as you will be able to see them on mouse over and it will not affect many alleles/sessions.

**Comment by ValWood** [comment]
closing, since non of the community  curators seem to be confused by this
